Tensile/KernelWriter.py

```diff
@@ -119,74 +114,174 @@ def makeSchedule(self, kernel, tensorParametersA, tensorParametersB, localWriteE
       # ds_read[A][1:]
       for i in range(self.numReadsPerIterA-self.numReadPerVectorA):
         latencyLeft -= tensorParametersA["localReadInstruction"].IssueLatency*2
         if latencyLeft < 0:
-          self.numMfmaForNextLoopLR += 1
-          latencyLeft = self.miLatency - tensorParametersA["localReadInstruction"].IssueLatency*2
+          self.numMfmaForLR += 1
+          latencyLeft = max(self.miLatencyLeft - tensorParametersA["localReadInstruction"].IssueLatency*2,0)
       # ds_read[B][1:]
       for i in range(self.numReadsPerIterB-self.numReadPerVectorB):
         latencyLeft -= tensorParametersB["localReadInstruction"].IssueLatency*2
         if latencyLeft < 0:
-          self.numMfmaForNextLoopLR += 1
-          latencyLeft = self.miLatency - tensorParametersB["localReadInstruction"].IssueLatency*2
+          self.numMfmaForLR += 1
+          latencyLeft = max(self.miLatencyLeft - tensorParametersB["localReadInstruction"].IssueLatency*2,0)
       # to calculate number of mfma we need to wait before data arrive from lds to vgpr.
-      # TODO: latency: 40 quad-cycle for 4 word, 22 quad-cycle for 2 word, 10 quad-cycle for 1 word
-      latencyForLR = tensorParametersB["localReadInstruction"].IssueLatency*18
-      latencyForLR -= latencyLeft # remaining latency in mfma
-      latencyForLR -= (self.miLatency+self.miLatencyBuffer+2) # last LR will have 1 mfma latency
+      # latency: 40 quad-cycle for 4 word, 20 quad-cycle for 2 word, 10 quad-cycle for 1 word / half word
+      self.numMfmaForNextLoopLR = self.numMfmaForLR
+      latencyForLR = roundUp(tensorParametersB["localReadInstruction"].blockWidth)*10
+      latencyForLR -= max(latencyLeft,0) # remaining latency in mfma
+      latencyForLR -= self.miLatency # last LR will have 1 mfma latency
       while latencyForLR > 0:
-        latencyForLR -= (self.miLatency+self.miLatencyBuffer+2)
+        latencyForLR -= self.miLatency
         self.numMfmaForNextLoopLR += 1
       # final index definition
       self.numMfmaForNextLoopLR = min(self.numMfmaForNextLoopLR,numMfmaPerIter-1)
       self.barrierMfmaIndex = numMfmaPerIter*(kernel["LoopIters"]-self.numItersPLR+1) - self.numMfmaForNextLoopLR - 1 if self.numItersPLR else 0
-      self.nextLoopLRMfmaIndex = self.barrierMfmaIndex if self.numItersPLR else numMfmaPerIter*kernel["LoopIters"] - self.numMfmaForNextLoopLR - 1
+      numMfmaBetweenLWandBarrier = 2 if kernel["MatrixInstM"] == 32 else 3
+      self.lwEndMfmaIndex = max(self.barrierMfmaIndex - numMfmaBetweenLWandBarrier,0) if self.numItersPLR else numMfmaPerIter*kernel["LoopIters"] - 1
+
+      #########
+      # Internally assign an optimized LWPM value for PGR2
+      #########
+      # strategy is to distribute LW/GR as wide as possible to avoid hitting vmem FIFO
+      # LWPM = (LW_End - LW_Start) / numLW
+      if kernel["LocalWritePerMfma"] == -3 and kernel["PrefetchGlobalRead"] == 2:
```

Comment:
What's the difference between LocalWritePerMfma -1, -2, -3?

Reply:
-1 is assign optimized LWPM externally (assign in SolutionStructs.py and override LWPM value)
-3 is assign optimized LWPM internally
-2 is used to support current logic yaml. If we set default to -1 or -3, some of configs in logic yaml will be rejected.